src/pint/models/binary_piecewise.py

```diff
@@ -0,0 +1,465 @@
+            )
+        for index in indices:
+            index_rf = f"{int(index):04d}"
+            for prefix in ["T0X_", "A1X_", "PLB_", "PUB_"]:
+                self.remove_param(prefix + index_rf)
+        self.validate()
+        self.setup()
+
+    def add_piecewise_param(self, param, j, param_unit=None, paramx=None):
+        """Add an orbital piecewise parameter.
+        Parameters
+        ----------
+        param : str
+                Piecewise parameter label e.g. "T0" or "A1".
+        param_unit : astropy.unit
+                Piecewise parameter unit.
+        paramx : np.float128
+                Piecewise parameter value.
+        j : int
+                Number to label the piece being added. Expected to match a set of piece boundaries.
+        """
+        if j is None:
+            dct = self.get_prefix_mapping_component(param + "X_")
+            j = np.max(list(dct.keys())) + 1
+        elif int(j) in self.get_prefix_mapping_component("X_"):
+            raise ValueError(
+                "Index '%s' is already in use in this model. Please choose another." % j
+            )
+        i = f"{int(j):04d}"
+
+        # handling if None are passed as arguments
+        if any(i is None for i in [param, param_unit, paramx]):
+            if param is not None:
+                # if parameter value or unit unset, set with default according to param
+                if param_unit is None:
+                    param_unit = (getattr(self, param)).units
+                if paramx is None:
+                    paramx = (getattr(self, param)).value
+        # check if name exits and is currently available
+        if param is None or param not in ["T0", "A1"]:
+            raise AttributeError(
+                "param '%s' not found. Please choose another. Currently implemented: 'T0' or 'A1' "
+                % param
+            )
+
+        if param == "A1":
+            self.add_param(
+                prefixParameter(
+                    name=param + f"X_{i}",
+                    units=param_unit,
+                    value=paramx,
+                    unit_template=lambda x: param_unit,
+                    description="Parameter" + param + "variation",
+                    description_template=lambda x: param,
+                    parameter_type="float",
+                    frozen=False,
+                )
+            )
+        elif param == "T0":
+            self.add_param(
+                prefixParameter(
+                    name=param + f"X_{i}",
+                    units=param_unit,
+                    value=paramx,
+                    unit_template=lambda x: param_unit,
+                    description="Parameter" + param + "variation",
+                    description_template=lambda x: param,
+                    parameter_type="float",
+                    frozen=False,
+                )
+            )
+
+    def lock_groups(self):
```

Comment:
Is this needed?  Is it used?

Reply:
So I found myself needing this function as validates/setups cannot be run before all pieces and corresponding group boundaries are declared. Otherwise there are floating piecewise parameters that don't have a group to go with them (which validate does not like). The function is intended to be be used after all add_group_range and add_piecewise_param have been called